src/plugins/legacy/itkDataImage/writers/itkDicomDataImageWriter.cpp

```diff
@@ -473,61 +485,84 @@ bool itkDicomDataImageWriter::write(const QString &path)
     if (!this->data())
         return false;
 
-    QString id = data()->identifier() ;
+    QString id = data()->identifier();
 
-    try {
-        if ( id == "itkDataImageChar3" )
-        {
-            writeDicom<char>(path);
-        }
-        else if ( id == "itkDataImageUChar3" )
-        {
-            writeDicom<unsigned char>(path);
-        }
-        else if ( id == "itkDataImageShort3" )
-        {
-            writeDicom<short>(path);
-        }
-        else if ( id == "itkDataImageUShort3" )
-        {
-            writeDicom<unsigned short>(path);
-        }
-        else if ( id == "itkDataImageInt3" )
+        try
         {
-            writeDicom<int>(path);
-        }
-        else if ( id == "itkDataImageUInt3" )
-        {
-            writeDicom<unsigned int>(path);
-        }
-        else if ( id == "itkDataImageLong3" )
-        {
-            writeDicom<long>(path);
-        }
-        else if ( id== "itkDataImageULong3" )
-        {
-            writeDicom<unsigned long>(path);
-        }
-        else if ( id == "itkDataImageFloat3" )
-        {
-            writeDicom<float>(path);
-        }
-        else if ( id == "itkDataImageDouble3" )
-        {
-            writeDicom<double>(path);
+            // Create a copy of data which can be modified if data is of type double, float, unsigned long or long
+            dataCopy = medAbstractDataFactory::instance()->createSmartPointer(id);
+            if (!dataCopy)
+            {
+                qWarning() << "Pixel type not supported by the software";
+                return false;
+            }
+            dataCopy->setData(data()->data());
+
+            if (id == "itkDataImageChar3")
+            {
+                writeDicom<char>(path);
+            }
+            else if (id == "itkDataImageUChar3")
+            {
+                writeDicom<unsigned char>(path);
+            }
+            else if (id == "itkDataImageShort3")
+            {
+                writeDicom<short>(path);
+            }
+            else if (id == "itkDataImageUShort3")
+            {
+                writeDicom<unsigned short>(path);
+            }
+            else if (id == "itkDataImageInt3")
+            {
+                writeDicom<int>(path);
+            }
+            else if (id == "itkDataImageUInt3")
+            {
+                writeDicom<unsigned int>(path);
+            }
+            else
+            {
+                typedef itk::Image<int, 3> ImageTypeOutput;
+                dataCopy = medAbstractDataFactory::instance()->createSmartPointer(medUtilitiesITK::itkDataImageId<ImageTypeOutput>());
+                if (id == "itkDataImageLong3")
+                {
+                    typedef itk::Image<long, 3> ImageTypeInput;
+                    dataCopy->setData(castToInt3<ImageTypeInput>(data()));
```

Comment:
it's just that line line (532) is the same as lines 537, 542 and 547. But you can keep them if you want or it can be:

```
                if (id == "itkDataImageLong3")
                {
                    typedef itk::Image<long, 3> ImageTypeInput;
                }
                else if (id == "itkDataImageULong3")
                {
                    typedef itk::Image<unsigned long, 3> ImageTypeInput;
                }
                else if (id == "itkDataImageFloat3")
                {
                    typedef itk::Image<float, 3> ImageTypeInput;
                }
                else if (id == "itkDataImageDouble3")
                {
                    typedef itk::Image<double, 3> ImageTypeInput;
                }
                else
                {
                    qWarning() << "Pixel type not yet supported";
                    return false;
                }
                dataCopy->setData(castToInt3<ImageTypeInput>(data()));
```
Or even with a `switch` condition

Reply:
I can not do that because ImageTypeImput is defined inside the if condition after the else it does not exist

Reply:
indeed sorry!